fedimint-bin-tests/src/federation.rs

```diff
@@ -0,0 +1,277 @@
+        let db = Database::new(db, module_decode_stubs());
+        let module_gens = ClientModuleGenRegistry::from(vec![
+            DynClientModuleGen::from(WalletClientGen),
+            DynClientModuleGen::from(MintClientGen),
+            DynClientModuleGen::from(LightningClientGen),
+        ]);
+        let client =
+            mint_client::Client::new(cfg.clone(), decoders, module_gens, db, Default::default())
+                .await;
+        Ok(Self {
+            client: Arc::new(client),
+            members,
+        })
+    }
+}
+
+#[derive(Clone)]
+pub struct Fedimintd {
+    bitcoind: Daemon<Bitcoin>,
+}
+
+impl Daemon<Fedimintd> {
+    pub async fn new(
+        process_mgr: &ProcessMgr,
+        bitcoind: Daemon<Bitcoin>,
+        id: usize,
+    ) -> Result<Self> {
+        let bin_dir = env::var("FM_BIN_DIR")?;
+        let cfg_dir = env::var("FM_CFG_DIR")?;
+
+        // spawn fedimintd
+        let mut fedimintd = tokio::process::Command::new(format!("{bin_dir}/fedimintd"));
+        fedimintd
+            // TODO: $FM_FEDIMINTD_DATA_DIR
+            .arg(format!("{cfg_dir}/server-{id}"))
+            .envs(fedimint_env(id)?);
+
+        info!("fedimintd started");
+        Self::start(
+            process_mgr,
+            "fedimintd-{id}",
+            fedimintd,
+            Fedimintd { bitcoind },
+        )
+        .await
+        // TODO: wait for federation to start
+    }
+
+    pub async fn cmd(&self) -> Command {
+        todo!()
+    }
+
+    pub async fn federation_id(&self) -> Result<String> {
+        todo!()
+    }
+
+    pub async fn await_gateway_registered(&self) -> Result<()> {
+        todo!()
+    }
+}
+
+pub async fn run_dkg(servers: usize) -> anyhow::Result<()> {
+    async fn create_tls(id: usize, sender: Sender<String>) -> anyhow::Result<()> {
+        // set env vars
+        let bin_dir = env::var("FM_BIN_DIR")?;
+        let server_name = format!("Server-{id}");
+        let env_vars = fedimint_env(id)?;
+        let p2p_url = env_vars.get("FM_P2P_URL").context("FM_P2P_URL not found")?;
+        let api_url = env_vars.get("FM_API_URL").context("FM_API_URL not found")?;
+        let out_dir = env_vars
+            .get("FM_FEDIMINT_DATA_DIR")
+            .context("FM_FEDIMINT_DATA_DIR not found")?;
+        let cert_path = format!("{out_dir}/tls-cert");
+
+        // create out-dir
+        fs::create_dir(&out_dir).await?;
+
+        info!("creating TLS certs created for started {server_name} in {out_dir}");
+        let mut task = tokio::process::Command::new(format!("{bin_dir}/distributedgen"))
+            .envs(fedimint_env(id)?)
+            .arg("create-cert")
+            .arg(format!("--p2p-url={p2p_url}"))
+            .arg(format!("--api-url={api_url}"))
+            .arg(format!("--out-dir={out_dir}"))
+            .arg(format!("--name={server_name}"))
+            .spawn()?;
+
+        // TODO: why is kill here?
+        // kill_on_exit(&task).await?;
```

Comment:
This just saves process IDs to a file which the shell scripts use to kill when they receive `EXIT` signal https://github.com/fedimint/fedimint/blob/5cf843dfe3ec9d3c80e153abf4e6168d3bfe26ee/scripts/build.sh#L93. I think this framework should be able to handle all cleanup on it's own.

Reply:
it is meant to be done by `struct ProcessHandle` (which is stored in `Daemon<...>`). if all copies `ProcessHandle` is dropped, the daemon will be killed. 

at high level we need to make sure that drop handlers are run correct, which should be the default in case of normal shutdown. and in `Ctrl-C` we just need to kills the tokio runtime properly (which is easy using tokio::signal)